pkg/services/ngalert/store/instance_database.go

```diff
@@ -87,43 +88,49 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
-		if err != nil {
-			return err
-		}
-
-		alertInstance := &models.AlertInstance{
-			RuleOrgID:         cmd.RuleOrgID,
-			RuleUID:           cmd.RuleUID,
-			Labels:            cmd.Labels,
-			LabelsHash:        labelsHash,
-			CurrentState:      cmd.State,
-			CurrentReason:     cmd.StateReason,
-			CurrentStateSince: cmd.CurrentStateSince,
-			CurrentStateEnd:   cmd.CurrentStateEnd,
-			LastEvalTime:      cmd.LastEvalTime,
-		}
-
-		if err := models.ValidateAlertInstance(alertInstance); err != nil {
-			return err
-		}
-
-		params := append(make([]interface{}, 0), alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash, alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(), alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
-
+// SaveAlertInstances saves all the provided alert instances to the store. It
+// writes transactions with a maximum count of 400 rows. Larger writes are
+// split into multiple transactions.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	// SQLite has a max 999 variable limit per statement. We use almost 9 per
+	// row, so we split our writes into fewer prepared statements using the row
+	// limit.
+	err := st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
```

Comment:
This does not start the transaction unless it is running in an existing transaction 
https://github.com/grafana/grafana/blob/6c43eb0b4dc876d2cda7d95ec229e66b4f0ceaec/pkg/services/sqlstore/session.go#L73

Reply:
My read of this is:
1. If there's an open transaction, this call reuses it.
2. If not, this call creates a new transaction.

That seems like reasonable behavior to me. What's better about using `WithTransactionalDbSession`? That guarantees the transaction is isolated, but it also means we can't compose these calls like delete or write with other SQL writes in a single transaction.